Applications/spire/source/ui/toggle_button.cpp

```diff
@@ -0,0 +1,119 @@
+#include "spire/ui/toggle_button.hpp"
+#include <QEvent>
+#include <QFocusEvent>
+#include <QHBoxLayout>
+#include <QMouseEvent>
+#include <QPainter>
+#include <QPaintEvent>
+#include "spire/spire/dimensions.hpp"
+
+using namespace boost::signals2;
+using namespace Spire;
+
+ToggleButton::ToggleButton(QImage icon, QImage toggle_icon, QWidget* parent)
+    : ToggleButton(icon, toggle_icon, icon, parent) {}
+
+ToggleButton::ToggleButton(QImage icon, QImage toggle_icon, QImage hover_icon,
+    QWidget* parent)
+    : ToggleButton(icon, toggle_icon, hover_icon, icon, parent) {}
+
+ToggleButton::ToggleButton(QImage icon, QImage toggle_icon, QImage hover_icon,
+    QImage disable_icon, QWidget* parent)
+    : QWidget(parent),
+      m_is_focused(false),
+      m_icon_button(icon, hover_icon, icon,
+        parent),
+      m_icon(icon),
+      m_hover_icon(hover_icon),
+      m_toggle_icon(std::move(toggle_icon)),
+      m_disabled_icon(std::move(disable_icon)) {
+  setFocusPolicy(Qt::StrongFocus);
+  m_icon_button.installEventFilter(this);
+  m_icon_button.setFocusPolicy(Qt::NoFocus);
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins({});
+  layout->addWidget(&m_icon_button);
+}
+
+void ToggleButton::setEnabled(bool enabled) {
+  set_icons(enabled);
+  QWidget::setEnabled(enabled);
+}
+
+void ToggleButton::setDisabled(bool disabled) {
+  set_icons(!disabled);
+  QWidget::setDisabled(disabled);
+}
+
+bool ToggleButton::eventFilter(QObject* object, QEvent* event) {
+  if(event->type() == QEvent::WindowActivate ||
+      event->type() == QEvent::WindowDeactivate) {
+    return true;
+  } else if(event->type() == QEvent::MouseButtonRelease) {
+    auto e = static_cast<QMouseEvent*>(event);
+    if(e->button() == Qt::LeftButton) {
+      swap_toggle();
+    }
+  }
+  return QWidget::eventFilter(object, event);
+}
+
+void ToggleButton::focusInEvent(QFocusEvent* event) {
+  if(event->reason() == Qt::TabFocusReason ||
+      event->reason() == Qt::BacktabFocusReason) {
+    m_is_focused = true;
+    update();
+  }
+}
+
+void ToggleButton::focusOutEvent(QFocusEvent* event) {
+  m_is_focused = false;
+  update();
+}
+
+void ToggleButton::keyPressEvent(QKeyEvent* event) {
+  if(event->key() == Qt::Key_Enter || event->key() == Qt::Key_Return ||
+      event->key() == Qt::Key_Space) {
+    swap_toggle();
+  } else {
+    parent()->event(event);
```

Comment:
I don't think this is valid in ```Qt```. The proper way to handle this is to ```reject``` the event or call your base classes ```keyPressEvent```.